File: frontend/src/pages/SaveAge/schemas.ts
import { z } from "zod";

export const DropdownListItemSchema = z.array(
  z.object({
    id: z.number(),
    name: z.string(),
  })
);

export const CensusDetailSchema = z.object({
  census_detail_id: z.number(),
  census_master_id: z.number(),
  birthdate: z.string(),
  relationship: z.string(),
  tobacco_disposition: z.string(),
  effective_date: z.string(),
});

export const CensusDetailListSchema = z.array(CensusDetailSchema);

export const CensusMasterSchema = z.object({
  census_master_id: z.number(),
  census_name: z.string(),
  census_path: z.string().optional(),
});

export type CensusMasterType = z.infer<typeof CensusMasterSchema>;

export const RateDetailSchema = z.object({
  rate_detail_id: z.number(),
  rate_master_id: z.number(),
  lower_age: z.number(),
  upper_age: z.number(),
  relationship: z.string(),
  tobacco_disposition: z.string(),
  rate: z.number(),
});

export const RateMasterSchema = z.object({
  rate_master_id: z.number(),
  rate_master_name: z.string(),
});

export type RateMasterType = z.infer<typeof RateMasterSchema>;

export const RateMasterWithDetailsSchema = z.object({
  rate_master_id: z.number(),
  rate_master_name: z.string(),
  rate_details: z.array(RateDetailSchema),
});

export const SaveAgeDataSchema = z.object({
  census_detail_id: z.number(),
  relationship: z.string(),
  tobacco_disposition: z.string(),
  issue_age: z.number(),
  birthdate: z.string(),
  save_age_effective_date: z.string(),
  new_effective_date: z.string(),
  save_age_rate: z.number().nullish(),
  new_rate: z.number().nullish(),
  diff: z.number().nullish(),
});

export type SaveAgeDataType = z.infer<typeof SaveAgeDataSchema>;

export const SaveAgeOutputSchema = z.object({
  data: z.array(SaveAgeDataSchema),
  stats: z.object({
    count: z.number(),
    save_age_rate: z.number().nullish(),
    new_rate: z.number().nullish(),
    diff: z.number().nullish(),
  }),
});

export const CensusStatsSchema = z.object({
  tobacco_stats: z.array(
    z.object({
      tobacco_disposition: z.string(),
      count: z.number(),
    })
  ),
  issue_age_stats: z.array(
    z.object({
      issue_age: z.number(),
      count: z.number(),
    })
  ),
  relationship_stats: z.array(
    z.object({
      relationship: z.string(),
      count: z.number(),
    })
  ),
  tenure_stats: z.array(
    z.object({
      tenure: z.number(),
      count: z.number(),
    })
  ),
});

export const ValidEffectiveDateSchema = z.string().date();

export type CensusStatsType = z.infer<typeof CensusStatsSchema>;

File: api/census/schemas.py
from extensions import ma
from shared import BaseSchema

from . import models as md


class SchemaCensusDetail(BaseSchema):
    class Meta:
        model = md.ModelCensusDetail
        load_instance = True
        include_relationships = True
        include_fk = True


class SchemaCensusMaster(BaseSchema):
    class Meta:
        model = md.ModelCensusMaster
        load_instance = True
        include_relationships = True
        include_fk = True

    census_details = ma.Nested(SchemaCensusDetail, many=True)


class SchemaCensusMasterDropdown(BaseSchema):
    census_master_id = ma.Integer(data_key="id")
    census_name = ma.String(data_key="name")


class SchemaCensusUpload(ma.Schema):
    tab = ma.String()
    birthdate = ma.Date()
    relationship = ma.String()
    tobacco_disposition = ma.String()
    effective_date = ma.Date()


class SchemaRateUpload(ma.Schema):
    rate_master_id = ma.Integer()
    lower_age = ma.Integer()
    upper_age = ma.Integer()
    relationship = ma.String()
    tobacco_disposition = ma.String()
    rate = ma.Float()


class SchemaRateDetail(BaseSchema):
    class Meta:
        model = md.ModelRateDetail
        load_instance = True
        include_relationships = True
        include_fk = True


class SchemaRateMaster(BaseSchema):
    class Meta:
        model = md.ModelRateMaster
        load_instance = True
        include_relationships = True
        include_fk = True

    rate_details = ma.Nested(SchemaRateDetail, many=True)


class SchemaRateMasterDropdown(BaseSchema):
    rate_master_id = ma.Integer(data_key="id")
    rate_master_name = ma.String(data_key="name")


class SchemaSaveAgeInputs(ma.Schema):
    effective_date = ma.Date(required=True)
    rate_master_id = ma.Integer(required=True)
    census_master_id = ma.Integer(required=True)


class SchemaSaveAgeOutput(ma.Schema):
    census_detail_id = ma.Integer()
    relationship = ma.String()
    tobacco_disposition = ma.String()
    issue_age = ma.Integer()
    birthdate = ma.Date("%Y-%m-%d")
    save_age_effective_date = ma.Date("%Y-%m-%d")
    new_effective_date = ma.Date("%Y-%m-%d")
    save_age_rate = ma.Float()
    new_rate = ma.Float()
    diff = ma.Float()


class SchemaCensusConfigLLM(ma.Schema):
    tab_name = ma.String()
    start_row_number = ma.Integer()
    start_column_number = ma.Integer()
    column_mapper = ma.Dict()
